Make TR_Uncopyable more widely available as TR::Uncopyable

jdmpapin opened the pull request:
`TR_Uncopyable` provides highly general functionality that should be available without including FrontEnd.hpp. This commit moves it into its own header, where it will be more widely available.

Additionally:

- It is moved into the `TR` namespace.

- The constructor and destructor are declared `protected` rather than `public`. Only subclasses need access to those.

- The only existing point of use (`TR_FrontEnd`) is changed to use `private` rather than `public` inheritance, since the fact that inheritance is used to get the desired effect is an implementation detail.

- A couple of types are changed to use `TR::Uncopyable`.

0xdaryl commented on compiler/infra/Uncopyable.hpp:
```diff
@@ -0,0 +1,40 @@
+#define OMR_UNCOPYABLE_INCL
+
+namespace TR {
+
```
It would be useful to have a brief comment here for what and why this class is useful and how it is expected to be used (i.e., subclassing).  Would you mind adding a brief comment here please?

0xdaryl commented on compiler/infra/Uncopyable.hpp:
```diff
@@ -0,0 +1,40 @@
+ *******************************************************************************/
+
+#ifndef OMR_UNCOPYABLE_INCL
+#define OMR_UNCOPYABLE_INCL
```
Technically this should be `TR_UNCOPYABLE_INCL` per existing conventions.

0xdaryl commented on compiler/infra/Uncopyable.hpp:
```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 IBM Corp. and others
```
As this was lifted from FrontEnd.hpp, I think the first copyright date should be 2000.

jdmpapin commented:
Updated to fix the copyright start date and include guard name, and to document the class. I also added another comment to explain how it works.

While documenting this, I realized that it additionally happens to prevent the implicit default definition of the move constructor and move assignment operator (in both `Uncopyable` itself and then therefore in its subtypes). I think moves usually make sense for a type even if copies don't. This could easily be made movable and therefore allow its subtypes to implicitly define move operations like so:
```
Uncopyable(Uncopyable&&) {}
Uncopyable &operator=(Uncopyable&&) { return *this; }
```
However, it isn't clear that we have a reasonable level of support for move semantics in all of our build compilers (even though the [list of supported C++ features](https://github.com/eclipse/omr/blob/21811c62904983392c7d8060b07c2af4ff633429/doc/SupportedC++Features.md#supported-c11-features) claims that we can use it). For example, cppreference.com says that no version of XLC++ supports "defaulted move special member functions." From a bit of experimentation, it seems that this includes implicitly-defined move operations, which means that we couldn't portably rely on the implicit definitions allowed by the snippet above.

But it seems we don't use move semantics regardless. At least, I don't recall seeing any explicit rvalue references or moves in the codebase, and in my experimentation I hit a number of surprising road blocks beyond just the lack of defaulted move methods, e.g. lack of move constructors for STL types. So for now I've left the move constructor and assignment operator undeclared without so much as commenting about them in the code, just as though move semantics don't even exist.

0xdaryl commented:
Jenkins build all
